File: DRRR.Server/Services/SystemMessagesService.cs
using Microsoft.AspNetCore.Mvc;
using Microsoft.Extensions.Configuration;
using System;
using System.IO;
using System.Text.RegularExpressions;

namespace DRRR.Server.Services
{
    /// <summary>
    /// 提供系统通知消息的服务
    /// </summary>
    public class SystemMessagesService
    {
        /// <summary>
        /// 服务器端返回用消息配置信息
        /// </summary>
        private readonly IConfiguration _systemMessageSettings;

        public SystemMessagesService()
        {
            _systemMessageSettings = new ConfigurationBuilder()
                .SetBasePath(Path.Combine(AppContext.BaseDirectory, "Resources"))
                .AddJsonFile("system-messages.server.json", optional: false, reloadOnChange: true)
                .AddEnvironmentVariables().Build();
        }

        /// <summary>
        /// 获取服务器端系统消息
        /// </summary>
        /// <param name="msgId">消息ID</param>
        /// <param name="args">用于替换占位符的参数</param>
        /// <returns>服务器端系统消息</returns>
        public string GetMessage(string msgId, params string[] args)
        {
            string value = _systemMessageSettings[msgId];
            return Regex.Replace(value, @"{(\d)}", match => args[int.Parse(match.Groups[1].Value)]);
        }
    }
}

File: DRRR.Client/src/app/core/system-messages.service.ts
import { Injectable } from '@angular/core';
import { Http } from '@angular/http';

/**
 * 提供系统消息资源
 */
@Injectable()
export class SystemMessagesService {
  private messages: object;

  constructor(private http: Http) {
    this.http.get('/api/resources/system-messages')
      .subscribe(data => {
        this.messages = data.json();
      });
  }

  /**
   * 获取特定的系统消息
   * @param {string} msgId 信息代号
   * @param {string[]} args 替换占位符用的参数
   * @returns {string} 特定的系统消息
   */
  getMessage(msgId: string, ...args: string[]): string {
    return ((this.messages || {})[msgId] || '').replace(/{(\d)}/g, (_ , i) => args[i]);
  }
}

